src/Service/MagicLinks/MagicLinkService.cs

```diff
@@ -1,27 +1,89 @@
 using Passwordless.Common.Services.Mail;
+using Passwordless.Service.Helpers;
 using Passwordless.Service.MagicLinks.Models;
 using Passwordless.Service.Models;
+using Passwordless.Service.Storage.Ef;
 
 namespace Passwordless.Service.MagicLinks;
 
-public class MagicLinkService
+public class MagicLinkService(
+    TimeProvider timeProvider,
+    ITenantStorage tenantStorage,
+    IFido2Service fido2Service,
+    IMailProvider mailProvider)
 {
-    private readonly IFido2Service _fido2Service;
-    private readonly IMailProvider _mailProvider;
-
-    public MagicLinkService(IFido2Service fido2Service, IMailProvider mailProvider)
+    private async Task EnforceLimitsAsync(MagicLinkDTO dto)
     {
-        _fido2Service = fido2Service;
-        _mailProvider = mailProvider;
+        var now = timeProvider.GetUtcNow();
+        var account = await tenantStorage.GetAccountInformation();
+        var accountAge = now - account.CreatedAt;
+
+        // Applications created less than 24 hours ago can only send magic links to the admin email address
+        if (accountAge < TimeSpan.FromHours(24) &&
+            !account.AdminEmails.Contains(dto.EmailAddress.Address, StringComparer.OrdinalIgnoreCase))
+        {
+            throw new ApiException(
+                "Because your application has been created less than 24 hours ago, " +
+                "you can only request magic links to the admin email address.",
+                403
+            );
+        }
+
+        var maxMonthlyLimit = account.Features?.MagicLinkEmailMaxMonthlyLimit ?? 500;
+        var monthlyLimit = accountAge.TotalDays switch
+        {
+            // App created <24 hours ago
+            < 1 => 0.2 * maxMonthlyLimit,
+            // App created <3 days ago
+            < 3 => 0.5 * maxMonthlyLimit,
+            // App created <30 days ago
+            < 30 => 0.75 * maxMonthlyLimit,
+            // App created >30 days ago
+            _ => maxMonthlyLimit
+        };
+
+        if (await tenantStorage.GetDispatchedEmailCountAsync(TimeSpan.FromDays(30)) >=
+            (int)Math.Max(1, monthlyLimit))
+        {
+            throw new ApiException(
+                "You have reached your monthly quota for magic link emails. " +
+                "Please try again later.",
+                429
+            );
+        }
+
+        var maxMinutelyLimit = account.Features?.MagicLinkEmailMaxMinutelyLimit ?? 5;
+        var minutelyLimit = accountAge.TotalDays switch
+        {
+            // App created <24 hours ago
+            < 1 => 0.2 * maxMinutelyLimit,
+            // App created <3 days ago
+            < 3 => 0.5 * maxMinutelyLimit,
+            // App created <30 days ago
+            < 30 => 0.75 * maxMinutelyLimit,
+            // App created >30 days ago
+            _ => maxMinutelyLimit
+        };
+
+        if (await tenantStorage.GetDispatchedEmailCountAsync(TimeSpan.FromDays(30)) >=
```

Comment:
suggestion: `TimeSpan.FromDays(30)` should maybe be something like `TimeSpan.FromMinutes(5)`

If this is the "tempo" regulator (as in: you can't send all your emails in a 5 minute window), I'm not sure you want fetch all the 30 days?

Reply:
The 30 days is required for monthly quota